% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sc_workflow.R
\name{runscPipe}
\alias{runscPipe}
\title{runscPipe}
\usage{
runscPipe(sample_name, outdir, r1, r2 = NULL, read_structure, filter_settings,
  genome_index, exon_anno, fix_chr = FALSE, barcode_anno = NULL,
  max_mis = 1, UMI_cor = 1, gene_fl = FALSE, species, gene_id_type,
  nthreads)
}
\arguments{
\item{sample_name}{sample names, the name should not have any space.}

\item{outdir}{the output folder}

\item{r1}{file path of read1}

\item{r2}{file path of read2}

\item{read_structure}{a list contains read structure configuration. for more help see `?sc_trim_barcode`}

\item{filter_settings}{a list contains read filter settings for more help see `?sc_trim_barcode`}

\item{genome_index}{genome index used for \code{Rsubread::align}}

\item{exon_anno}{a vector of gff exon annotation file paths.}

\item{fix_chr}{whether add `chr` to chromosome names, fix inconsistant names between different annotations.}

\item{barcode_anno}{file path for cell barcode annotation.}

\item{max_mis}{maximum mismatch allowed in barcode. default to be 1}

\item{UMI_cor}{correct UMI sequence error: 0 means no correction, 1 means simple correction and merge UMI with distance 1.}

\item{gene_fl}{whether to remove low abundant gene count. low abundant is defined as only one copy of one UMI for this gene}

\item{species}{the organism of the data. List of possible names can be retrieved using the function 
`listDatasets`from `biomaRt` package. (i.e `mmusculus_gene_ensembl` or `hsapiens_gene_ensembl`)}

\item{gene_id_type}{gene id type of the data A possible list of ids can be retrieved using the function `listAttributes` from `biomaRt` package. 
the commonly used id types are `external_gene_name`, `ensembl_gene_id` or `entrezgene`}

\item{nthreads}{number of threads used}
}
\value{
an SCData object
}
\description{
runscPipe
}
