% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/wrapper_scPipeCPP.R
\name{sc_demultiplex}
\alias{sc_demultiplex}
\title{sc_demultiplex}
\usage{
sc_demultiplex(inbam, outdir, bc_anno, max_mis = 1, am = "YE", ge = "GE",
  bc = "YC", mb = "YM", mito = "MT")
}
\arguments{
\item{inbam}{input bam file. this should be the output of \code{sc_exon_mapping}}

\item{outdir}{output folder}

\item{bc_anno}{barcode annotation, first column is cell id, second column is cell barcode sequence}

\item{max_mis}{maximum mismatch allowed in barcode. default to be 1}

\item{am}{mapping status tag (default: YE)}

\item{ge}{gene id tag (default: GE)}

\item{bc}{cell barcode tag (default: YC)}

\item{mb}{molecular barcode tag (default: YM)}

\item{mito}{mitochondrial chromosome name, should be consistant with the chromosome name in bam file}
}
\description{
separate bam file by cell, output to outdir/count/[cell_id].csv.
the output contains information for all reads that can be mapped to exons.
including the gene id, UMI of that read and the distance to transcript end position.
}
\examples{
#TODO
}
